File: app/src/main/java/de/tum/in/tumcampusapp/cards/ChatMessagesCard.java
package de.tum.in.tumcampusapp.cards;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RemoteViews;

import com.google.gson.Gson;

import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.activities.ChatActivity;
import de.tum.in.tumcampusapp.auxiliary.Const;
import de.tum.in.tumcampusapp.cards.generic.Card;
import de.tum.in.tumcampusapp.cards.generic.NotificationAwareCard;
import de.tum.in.tumcampusapp.database.TcaDb;
import de.tum.in.tumcampusapp.database.dao.ChatMessageDao;
import de.tum.in.tumcampusapp.models.dbEntities.ChatRoomDbRow;
import de.tum.in.tumcampusapp.models.tumcabe.ChatMessage;
import de.tum.in.tumcampusapp.models.tumcabe.ChatRoom;

import static de.tum.in.tumcampusapp.managers.CardManager.CARD_CHAT;

/**
 * Card that shows the cafeteria menu
 */
public class ChatMessagesCard extends NotificationAwareCard {
    private List<ChatMessage> mUnread;
    private String mRoomName;
    private int mRoomId;
    private String mRoomIdString;
    private final ChatMessageDao chatMessageDao;


    public ChatMessagesCard(Context context, ChatRoomDbRow room) {
        super(CARD_CHAT, context, "card_chat");
        TcaDb tcaDb = TcaDb.getInstance(context);
        chatMessageDao = tcaDb.chatMessageDao();
        setChatRoom(room.getName(),room.getRoom(),room.getSemesterId() + ':' + room.getName());
    }

    public static Card.CardViewHolder inflateViewHolder(ViewGroup parent) {
        View view = LayoutInflater.from(parent.getContext())
                                  .inflate(R.layout.card_item, parent, false);
        return new Card.CardViewHolder(view);
    }

    @Override
    public String getTitle() {
        return mRoomName;
    }

    @Override
    public void updateViewHolder(RecyclerView.ViewHolder viewHolder) {
        mCard = viewHolder.itemView;
        CardViewHolder cardsViewHolder = (CardViewHolder) viewHolder;
        List<View> addedViews = cardsViewHolder.getAddedViews();

        mLinearLayout = mCard.findViewById(R.id.card_view);
        mTitleView = mCard.findViewById(R.id.card_title);
        mTitleView.setText(mRoomName);

        //Remove additional views
        for (View view : addedViews) {
            mLinearLayout.removeView(view);
        }

        // Show cafeteria menu
        for (ChatMessage message : mUnread) {
            addedViews.add(addTextView(message.getMember()
                                              .getDisplayName() + ": " + message.getText()));
        }
    }

    /**
     * Sets the information needed to build the card
     *
     * @param roomName Name of the chat room
     * @param roomId   Id of the chat room
     */
    public void setChatRoom(String roomName, int roomId, String roomIdString) {
        mRoomName = roomName;
        mRoomName = mRoomName.replaceAll("[A-Z, 0-9(LV\\.Nr)=]+$", "");
        mRoomName = mRoomName.replaceAll("\\([A-Z]+[0-9]+\\)", "");
        mRoomName = mRoomName.replaceAll("\\[[A-Z]+[0-9]+\\]", "");
        mRoomName = mRoomName.trim();
        chatMessageDao.deleteOldEntries();
        //mUnread = chatMessageDao.getLastUnread(roomId); TODO
        mRoomIdString = roomIdString;
        mRoomId = roomId;
    }

    @Override
    public Intent getIntent() {
        Bundle bundle = new Bundle();
        final Intent intent = new Intent(mContext, ChatActivity.class);
        ChatRoom currentChatRoom = new ChatRoom(mRoomIdString);
        currentChatRoom.setId(mRoomId);
        intent.putExtra(Const.CURRENT_CHAT_ROOM, new Gson().toJson(currentChatRoom));
        intent.putExtras(bundle);
        return intent;
    }

    @Override
    public int getId() {
        return mRoomId;
    }

    @Override
    protected void discard(Editor editor) {
        chatMessageDao.markAsRead(getId());
    }

    @Override
    protected boolean shouldShowNotification(SharedPreferences prefs) {
        return false;
    }

    @Override
    public RemoteViews getRemoteViews(Context context) {
        final RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.cards_widget_card);
        remoteViews.setTextViewText(R.id.widgetCardTextView, this.getTitle());
        remoteViews.setImageViewResource(R.id.widgetCardImageView, R.drawable.ic_comment);
        return remoteViews;
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/chat/ChatMessagesCard.kt
package de.tum.`in`.tumcampusapp.component.ui.chat

import android.content.Context
import android.content.SharedPreferences.Editor
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.other.navigation.NavDestination
import de.tum.`in`.tumcampusapp.component.ui.chat.activity.ChatActivity
import de.tum.`in`.tumcampusapp.component.ui.chat.model.ChatMessage
import de.tum.`in`.tumcampusapp.component.ui.chat.model.ChatRoom
import de.tum.`in`.tumcampusapp.component.ui.chat.model.ChatRoomDbRow
import de.tum.`in`.tumcampusapp.component.ui.overview.CardInteractionListener
import de.tum.`in`.tumcampusapp.component.ui.overview.CardManager.CARD_CHAT
import de.tum.`in`.tumcampusapp.component.ui.overview.card.Card
import de.tum.`in`.tumcampusapp.component.ui.overview.card.CardViewHolder
import de.tum.`in`.tumcampusapp.database.TcaDb
import de.tum.`in`.tumcampusapp.utils.Const
import java.util.ArrayList

/**
 * Card that shows the cafeteria menu
 */
class ChatMessagesCard(
    context: Context,
    room: ChatRoomDbRow
) : Card(CARD_CHAT, context, "card_chat") {

    private var mUnread: List<ChatMessage> = ArrayList()
    private var nrUnread = 0
    private var mRoomName = ""
    private var mRoomId = 0
    private var mRoomIdString = ""

    private val chatMessageDao: ChatMessageDao

    override val optionsMenuResId: Int
        get() = R.menu.card_popup_menu

    init {
        val tcaDb = TcaDb.getInstance(context)
        chatMessageDao = tcaDb.chatMessageDao()
        setChatRoom(room.name, room.room, "${room.semesterId}:${room.name}")
    }

    override fun updateViewHolder(viewHolder: RecyclerView.ViewHolder) {
        super.updateViewHolder(viewHolder)
        val chatMessagesViewHolder = viewHolder as? ChatMessagesCardViewHolder
        chatMessagesViewHolder?.bind(mRoomName, mRoomId, mRoomIdString, mUnread)
    }

    /**
     * Sets the information needed to build the card
     *
     * @param roomName Name of the chat room
     * @param roomId Id of the chat room
     */
    private fun setChatRoom(roomName: String, roomId: Int, roomIdString: String) {
        mRoomName = listOf("[A-Z, 0-9(LV\\.Nr)=]+$", "\\([A-Z]+[0-9]+\\)", "\\[[A-Z]+[0-9]+\\]")
                .map { it.toRegex() }
                .fold(roomName) { name, regex -> name.replace(regex, "") }
                .trim()
        chatMessageDao.deleteOldEntries()
        nrUnread = chatMessageDao.getNumberUnread(roomId)
        mUnread = chatMessageDao.getLastUnread(roomId).asReversed()
        mRoomIdString = roomIdString
        mRoomId = roomId
    }

    override fun getNavigationDestination(): NavDestination? {
        val bundle = Bundle().apply {
            val chatRoom = ChatRoom(mRoomIdString).apply { id = mRoomId }
            val value = Gson().toJson(chatRoom)
            putString(Const.CURRENT_CHAT_ROOM, value)
        }
        return NavDestination.Activity(ChatActivity::class.java, bundle)
    }

    override fun getId() = mRoomId

    override fun discard(editor: Editor) = chatMessageDao.markAsRead(mRoomId)

    companion object {

        @JvmStatic
        fun inflateViewHolder(
            parent: ViewGroup,
            interactionListener: CardInteractionListener
        ): CardViewHolder {
            val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.card_chat_messages, parent, false)
            return ChatMessagesCardViewHolder(view, interactionListener)
        }
    }
}
